File: engine/wolves/snapshot.py
from __future__ import annotations

from pydantic import BaseModel, ConfigDict, Field

SCHEMA_VERSION = 1


class Candidate(BaseModel):
    team_id: str
    prob: float


class SlotSide(BaseModel):
    label: str
    candidates: list[Candidate]


class Slot(BaseModel):
    match: int
    stage: str
    date: str
    city: str
    home: SlotSide
    away: SlotSide


class RoundOpponents(BaseModel):
    """Opponent distribution for a later round, conditional on the group finish and the focus team reaching it."""

    round: str
    match: int
    city: str
    date: str
    reach_prob: float
    opponents: list[Candidate]


class FocusTeamPath(BaseModel):
    finish: str
    prob: float
    r32_match: int
    city: str
    date: str
    opponents: list[Candidate]
    onward: list[RoundOpponents] = Field(default_factory=list)


class ModalStep(BaseModel):
    round: str
    match: int
    city: str
    date: str
    opponent_id: str
    opponent_prob: float


class CityProb(BaseModel):
    city: str
    prob: float


class LockDate(BaseModel):
    """How certain the focus team's R32 city is once a group matchday completes; bookability signal."""

    date: str
    prob_locked: float
    locked_city_probs: dict[str, float]


class WhatIfOutcome(BaseModel):
    outcome: str
    prob: float
    finish_probs: dict[str, float]
    r32_city_probs: dict[str, float]


class WhatIfFixture(BaseModel):
    match: int
    date: str
    city: str
    opponent_id: str
    outcomes: list[WhatIfOutcome]


class FocusTeamBlock(BaseModel):
    team_id: str
    group: str
    finish_probs: dict[str, float]
    reach_probs: dict[str, float]
    paths: list[FocusTeamPath]
    modal_path: list[ModalStep] = Field(default_factory=list)
    city_probs: dict[str, list[CityProb]] = Field(default_factory=dict)
    lock_dates: list[LockDate] = Field(default_factory=list)
    what_if: list[WhatIfFixture] = Field(default_factory=list)


class GroupTeamStanding(BaseModel):
    team_id: str
    finish_probs: dict[str, float]
    expected_points: float


class GroupBlock(BaseModel):
    group: str
    teams: list[GroupTeamStanding]


class MatchProbs(BaseModel):
    """Forecast for one unplayed fixture. Group probabilities are within 90
    minutes; knockout figures are conditional on the modal pairing occurring."""

    match: int
    stage: str
    date: str
    city: str
    home_id: str
    away_id: str
    p_home: float
    p_away: float
    p_draw: float | None = None
    p_decided_90: float | None = None
    p_pairing: float | None = None
    modal_score: str | None = None


class ResultSetEntry(BaseModel):
    match: int
    home_id: str | None = None
    away_id: str | None = None
    home_goals: int
    away_goals: int
    winner: str | None = None
    source_fixture_id: int | None = None
    fetched_at: str | None = None


class ResultSetBlock(BaseModel):
    digest: str = ""
    results: list[ResultSetEntry] = Field(default_factory=list)


class RunMeta(BaseModel):
    run_id: str
    created_at: str
    as_of: str = ""
    n_sims: int
    engine_version: str
    kind: str


def run_day(meta: RunMeta) -> str:
    """The forecast day a run speaks for; runs are as_of-keyed, and the
    wall-clock created_at only stands in for replays of older snapshots."""
    return meta.as_of or meta.created_at[:10]


class TeamStoryOut(BaseModel):
    summary: str
    why: str


class NarrativeBlock(BaseModel):
    headline: str = ""
    team_stories: dict[str, TeamStoryOut] = Field(default_factory=dict)
    focus_story: str | None = None
    slot_rationales: dict[str, str] = Field(default_factory=dict)
    travel_memo: str | None = None


class LedgerEntryOut(BaseModel):
    id: str
    claim: str
    source_url: str
    # Joined from the article cache at snapshot build so the frontend can
    # name sources without exposing internal ledger machinery.
    title: str | None = None
    status: str
    mechanism: str
    proposed_delta: float = 0.0
    expiry: str | None = None
    team_id: str | None = None
    relevance: float | None = None
    source_tier: int | None = None
    retrieved_at: str | None = None
    retrieval_id: str | None = None
    created_at: str


class SourceRelevanceOut(BaseModel):
    """One source the agent ranked, fetched or cited during the run."""

    url: str
    title: str = ""
    hostname: str = ""
    tier: int | None = None
    score: float | None = None
    reason: str = ""
    sub_question: str = ""
    ranked: bool = False
    cited: bool = False
    fetched: bool = False
    seen_in_run: str | None = None
    retrieval_id: str | None = None
    created_by: str = ""


class ScenarioWeightOut(BaseModel):
    name: str
    weight: float
    scenario_id: str | None = None
    ledger_ids: list[str] = Field(default_factory=list)
    rationale: str = ""
    camp: str = ""
    label: str = ""
    summary: str = ""


class CampOut(BaseModel):
    key: str
    label: str = ""
    summary: str = ""
    weight: float = 0.0
    order: int = 0


class MarketGapOut(BaseModel):
    """A published market stance with separate model and forecast comparisons."""

    team_id: str
    model_prob: float
    market_prob: float
    forecast_prob: float | None = None
    model_market_gap_pp: float | None = None
    forecast_market_gap_pp: float | None = None
    gap_pp: float
    floor_multiple: float | None = None
    direction: str
    model_direction: str = ""


class WorldOut(BaseModel):
    """One published world's configuration, kept so live republishes can
    reapply the agent's adjustments without re-running the agent."""

    name: str
    weight: float
    perturbations: list[dict] = Field(default_factory=list)
    latent_effects: list[dict] = Field(default_factory=list)
    title_probs: dict[str, float] = Field(default_factory=dict)
    # Match id -> {home, draw, away}; the surface the spread P&L is scored on.
    match_probs: dict[str, dict[str, float]] = Field(default_factory=dict)


class QuantFindingOut(BaseModel):
    """One quant node's headline, surfaced so the run page can show the analysis."""

    node_id: str
    summary: str
    headline_value: float | None = None
    findings: list[str] = Field(default_factory=list)


class GovernorOut(BaseModel):
    scale: float = 1.0
    effective_d: float = 1.0


class MovementOut(BaseModel):
    previous_prob: float
    current_prob: float
    delta_pp: float


class AttributionOut(BaseModel):
    bracket_pp: dict[str, float] = Field(default_factory=dict)
    refit_pp: dict[str, float] = Field(default_factory=dict)
    residual_pp: dict[str, float] = Field(default_factory=dict)
    movement: dict[str, MovementOut] = Field(default_factory=dict)


class CalibrationSummary(BaseModel):
    matches_scored: int = 0
    brier: dict[str, float] = Field(default_factory=dict)
    log_loss: dict[str, float] = Field(default_factory=dict)
    adjustment_pnl: float | None = None
    governor_scale: float = 1.0
    spread_pnl: float | None = None
    band_coverage: float | None = None
    movement_ratio: float | None = None


class ProvenanceOut(BaseModel):
    news_considered: int = 0
    news_material: int = 0
    news_excluded: int = 0
    market_disagreements: int = 0
    noise_floor_pp: float = 0.0
    n_worlds: int = 1
    n_camps: int = 1


class RevisionOut(BaseModel):
    """Post-acceptance revision trace for later counterfactual scoring."""

    revisions_used: int = 0
    counterfactual_artifact_id: str = ""
    revision_rationale: str = ""


class FinalisationOut(BaseModel):
    artifact_id: str
    submission_fingerprint: str
    validation_issue_counts: dict[str, int] = Field(default_factory=dict)
    referee_status: str
    referee_reason: str = ""
    advertised_ceiling_usd: float
    forecast_reserved_usd: float
    referee_reserved_usd: float
    settled_cost_usd: float


class AgentBlock(BaseModel):
    """Agent-run extras; absent on sim-only snapshots. Additive by design."""

    narrative: NarrativeBlock
    artifact_id: str = ""
    ledger_entries: list[LedgerEntryOut] = Field(default_factory=list)
    sources: list[SourceRelevanceOut] = Field(default_factory=list)
    scenario_weights: list[ScenarioWeightOut] = Field(default_factory=list)
    camps: list[CampOut] = Field(default_factory=list)
    worlds: list[WorldOut] = Field(default_factory=list)
    quant_findings: list[QuantFindingOut] = Field(default_factory=list)
    escalations: list[str] = Field(default_factory=list)
    market_gaps: list[MarketGapOut] = Field(default_factory=list)
    market_justification: str = ""
    change_justification: str = ""
    inconsistency_note: str = ""
    news_impacts: dict[str, str] = Field(default_factory=dict)
    copy_guard_version: int | None = None
    attribution: AttributionOut | None = None
    governor: GovernorOut | None = None
    calibration: CalibrationSummary | None = None
    provenance: ProvenanceOut | None = None
    branch_audit: dict[str, object] | None = None
    world_metadata: dict[str, dict[str, object]] = Field(default_factory=dict)
    revision: RevisionOut | None = None
    finalisation: FinalisationOut | None = None


class ChampionBlock(BaseModel):
    """Which trusted model produced the simulation numbers."""

    id: str
    version: str
    dataset_id: str
    half_life_days: float | None = None
    blend_weight: float = 0.0
    # Played results overlaid into the strength refit.
    results_overlaid: int = 0


class TeamInterval(BaseModel):
    team_id: str
    lo: float
    hi: float


class TeamDistributions(BaseModel):
    """Per-stage epistemic spread for one team: open cells carry the quantile
    vector, settled cells carry the outcome as a flag, never both."""

    quantiles: dict[str, list[float]] = Field(default_factory=dict)
    settled: dict[str, int] = Field(default_factory=dict)


class NewsItemOut(BaseModel):
    """One sourced news item joined to its price; impact is the agent's why."""

    ledger_id: str
    claim: str
    mechanism: str
    source_url: str
    title: str | None = None
    hostname: str = ""
    status: str = ""
    signed_delta_pp: float | None = None
    material: bool = False
    excluded_reason: str | None = None
    impact: str | None = None


class TeamDriver(BaseModel):
    """Per-camp chances, any market stance, sourced news and disagreement shape for one team."""

    camp_probs: dict[str, float] = Field(default_factory=dict)
    market_gap: MarketGapOut | None = None
    news: list[NewsItemOut] = Field(default_factory=list)
    has_story: bool = False
    higher_camp: str | None = None
    spread_pp: float = 0.0
    noise_floor_pp: float = 0.0


class DistributionsBlock(BaseModel):
    """Confidence in each published number: weighted (world x parameter-draw)
    quantiles per team per stage. The headline stays the mean; this block is
    epistemic dispersion ("how settled the number is"), never the aleatory
    outcome range ("the tournament could still go any way"), which is a
    separate named quantity."""

    quantile_levels: list[float] = Field(default_factory=list)
    provenance: str = "parameters_only"
    n_worlds: int = 1
    width_floored: bool = False
    sidecar: str = ""
    teams: dict[str, TeamDistributions] = Field(default_factory=dict)
    drivers: dict[str, TeamDriver] = Field(default_factory=dict)


class MarketsBlock(BaseModel):
    """Transparency block: model view, de-vigged market and a reference blend.

    On agent runs teams[] is the published headline and blend_probs is
    comparison only; deterministic runs publish blend_probs as teams[]."""

    model_config = ConfigDict(protected_namespaces=())

    model_probs: dict[str, float] = Field(default_factory=dict)
    market_probs: dict[str, float] = Field(default_factory=dict)
    blend_probs: dict[str, float] = Field(default_factory=dict)
    model_weight: float = 0.0


class TeamInfo(BaseModel):
    team_id: str
    name: str
    group: str
    elo: float
    rating: float = 0.0
    value_eur_m: float | None = None
    champion_prob: float = 0.0
    reach_probs: dict[str, float] = Field(default_factory=dict)


class Snapshot(BaseModel):
    """The engine-to-web contract; additive changes only, breaking changes bump SCHEMA_VERSION."""

    schema_version: int = SCHEMA_VERSION
    run: RunMeta
    focus: FocusTeamBlock
    slots: list[Slot]
    teams: list[TeamInfo]
    groups: list[GroupBlock] = Field(default_factory=list)
    matches: list[MatchProbs] = Field(default_factory=list)
    agent: AgentBlock | None = None
    champion: ChampionBlock | None = None
    intervals: list[TeamInterval] = Field(default_factory=list)
    markets: MarketsBlock | None = None
    distributions: DistributionsBlock | None = None
    result_set: ResultSetBlock = Field(default_factory=ResultSetBlock)

File: web/src/lib/snapshot.ts
export interface Candidate {
  team_id: string;
  prob: number;
}

export interface SlotSide {
  label: string;
  candidates: Candidate[];
}

export interface Slot {
  match: number;
  stage: string;
  date: string;
  city: string;
  home: SlotSide;
  away: SlotSide;
}

export type Finish = "win_group" | "runner_up" | "third";

export interface RoundOpponents {
  round: string;
  match: number;
  city: string;
  date: string;
  reach_prob: number;
  opponents: Candidate[];
}

export interface FocusTeamPath {
  finish: Finish;
  prob: number;
  r32_match: number;
  city: string;
  date: string;
  opponents: Candidate[];
  onward?: RoundOpponents[];
}

export interface ModalStep {
  round: string;
  match: number;
  city: string;
  date: string;
  opponent_id: string;
  opponent_prob: number;
}

export interface CityProb {
  city: string;
  prob: number;
}

export interface LockDate {
  date: string;
  prob_locked: number;
  locked_city_probs: Record<string, number>;
}

export interface WhatIfOutcome {
  outcome: string;
  prob: number;
  finish_probs: Record<string, number>;
  r32_city_probs: Record<string, number>;
}

export interface WhatIfFixture {
  match: number;
  date: string;
  city: string;
  opponent_id: string;
  outcomes: WhatIfOutcome[];
}

export interface FocusTeamBlock {
  team_id: string;
  group: string;
  finish_probs: Record<string, number>;
  reach_probs: Record<string, number>;
  paths: FocusTeamPath[];
  modal_path?: ModalStep[];
  city_probs?: Record<string, CityProb[]>;
  lock_dates?: LockDate[];
  what_if?: WhatIfFixture[];
}

export interface RunMeta {
  run_id: string;
  created_at: string;
  as_of?: string;
  n_sims: number;
  engine_version: string;
  kind: string;
}

export interface TeamInfo {
  team_id: string;
  name: string;
  group: string;
  elo: number;
  rating?: number;
  value_eur_m?: number | null;
  champion_prob?: number;
  reach_probs?: Record<string, number>;
}

export interface GroupTeamStanding {
  team_id: string;
  finish_probs: Record<string, number>;
  expected_points: number;
}

export interface GroupBlock {
  group: string;
  teams: GroupTeamStanding[];
}

export interface MatchProbs {
  match: number;
  stage: string;
  date: string;
  city: string;
  home_id: string;
  away_id: string;
  p_home: number;
  p_away: number;
  p_draw?: number | null;
  p_decided_90?: number | null;
  p_pairing?: number | null;
  modal_score?: string | null;
}

export interface ResultSetEntry {
  match: number;
  home_id?: string | null;
  away_id?: string | null;
  home_goals: number;
  away_goals: number;
  winner?: string | null;
  source_fixture_id?: number | null;
  fetched_at?: string | null;
}

export interface ResultSetBlock {
  digest: string;
  results: ResultSetEntry[];
}

export interface TeamStoryOut {
  summary: string;
  why: string;
}

export interface NarrativeBlock {
  headline?: string;
  team_stories?: Record<string, TeamStoryOut>;
}

export interface LedgerEntryOut {
  id: string;
  claim: string;
  source_url: string;
  title: string | null;
  status: string;
  mechanism: string;
  proposed_delta: number;
  expiry: string | null;
  team_id: string | null;
  relevance: number | null;
  source_tier: number | null;
  retrieved_at: string | null;
  retrieval_id: string | null;
  created_at: string;
}

export interface SourceRelevanceOut {
  url: string;
  title?: string;
  hostname?: string;
  tier?: number | null;
  score?: number | null;
  reason?: string;
  sub_question?: string;
  ranked?: boolean;
  cited?: boolean;
  fetched?: boolean;
  seen_in_run?: string | null;
  retrieval_id?: string | null;
  created_by?: string;
}

export interface ScenarioWeightOut {
  name: string;
  weight: number;
  scenario_id: string | null;
  ledger_ids: string[];
  rationale?: string;
  camp?: string;
  label?: string;
  summary?: string;
}

export interface CampOut {
  key: string;
  label?: string;
  summary?: string;
  weight?: number;
  order?: number;
}

export interface MarketGapOut {
  team_id: string;
  model_prob: number;
  market_prob: number;
  forecast_prob?: number | null;
  model_market_gap_pp?: number | null;
  forecast_market_gap_pp?: number | null;
  gap_pp: number;
  floor_multiple: number | null;
  direction: string;
  model_direction: string;
}

export interface WorldOut {
  name: string;
  weight: number;
  perturbations: Record<string, unknown>[];
  latent_effects: Record<string, unknown>[];
  title_probs?: Record<string, number>;
  match_probs?: Record<string, Record<string, number>>;
}

export interface QuantFindingOut {
  node_id: string;
  summary: string;
  headline_value: number | null;
  findings: string[];
}

export interface GovernorOut {
  scale: number;
  effective_d: number;
}

export interface AttributionOut {
  bracket_pp: Record<string, number>;
  refit_pp: Record<string, number>;
  residual_pp: Record<string, number>;
  movement?: Record<string, { previous_prob: number; current_prob: number; delta_pp: number }>;
}

export interface FinalisationOut {
  artifact_id?: string;
  submission_fingerprint?: string;
  validation_issue_counts?: Record<string, number>;
  referee_status?: string;
  referee_reason?: string;
  advertised_ceiling_usd?: number;
  forecast_reserved_usd?: number;
  referee_reserved_usd?: number;
  settled_cost_usd?: number;
}

export interface CalibrationSummary {
  matches_scored: number;
  brier: Record<string, number>;
  log_loss: Record<string, number>;
  adjustment_pnl: number | null;
  governor_scale: number;
  spread_pnl?: number | null;
  band_coverage?: number | null;
  movement_ratio?: number | null;
}

export interface ProvenanceOut {
  news_considered: number;
  news_material: number;
  news_excluded: number;
  market_disagreements: number;
  noise_floor_pp: number;
  n_worlds: number;
  n_camps: number;
}

export interface RevisionOut {
  revisions_used: number;
  counterfactual_artifact_id: string;
  revision_rationale: string;
}

export interface AgentBlock {
  narrative: NarrativeBlock;
  artifact_id: string;
  ledger_entries: LedgerEntryOut[];
  sources?: SourceRelevanceOut[];
  scenario_weights: ScenarioWeightOut[];
  camps?: CampOut[];
  worlds: WorldOut[];
  quant_findings?: QuantFindingOut[];
  escalations: string[];
  market_gaps?: MarketGapOut[];
  market_justification: string;
  change_justification: string;
  inconsistency_note: string;
  news_impacts?: Record<string, string>;
  copy_guard_version?: number | null;
  attribution: AttributionOut | null;
  finalisation?: FinalisationOut | null;
  governor: GovernorOut | null;
  calibration: CalibrationSummary | null;
  provenance?: ProvenanceOut | null;
  branch_audit?: Record<string, unknown> | null;
  world_metadata?: Record<string, Record<string, unknown>>;
  revision?: RevisionOut | null;
}

export interface ChampionBlock {
  id: string;
  version: string;
  dataset_id: string;
  half_life_days?: number | null;
  blend_weight?: number;
  results_overlaid?: number;
}

export interface TeamInterval {
  team_id: string;
  lo: number;
  hi: number;
}

export interface TeamDistributions {
  quantiles: Record<string, number[]>;
  settled: Record<string, number>;
}

export interface NewsItemOut {
  ledger_id: string;
  claim: string;
  mechanism: string;
  source_url: string;
  title: string | null;
  hostname: string;
  status: string;
  signed_delta_pp: number | null;
  material: boolean;
  excluded_reason: string | null;
  impact: string | null;
}

export interface TeamDriver {
  camp_probs: Record<string, number>;
  market_gap: MarketGapOut | null;
  news: NewsItemOut[];
  has_story: boolean;
  higher_camp: string | null;
  spread_pp: number;
  noise_floor_pp: number;
}

export interface DistributionsBlock {
  quantile_levels: number[];
  provenance: string;
  n_worlds: number;
  width_floored: boolean;
  sidecar: string;
  teams: Record<string, TeamDistributions>;
  drivers?: Record<string, TeamDriver>;
}

export interface MarketsBlock {
  model_probs?: Record<string, number>;
  market_probs?: Record<string, number>;
  blend_probs?: Record<string, number>;
  model_weight?: number;
}

export interface Snapshot {
  schema_version: number;
  run: RunMeta;
  focus: FocusTeamBlock;
  slots: Slot[];
  teams: TeamInfo[];
  groups?: GroupBlock[];
  matches?: MatchProbs[];
  agent?: AgentBlock | null;
  champion?: ChampionBlock | null;
  intervals?: TeamInterval[];
  markets?: MarketsBlock | null;
  distributions?: DistributionsBlock | null;
  result_set?: ResultSetBlock;
}

export function teamNames(snapshot: Snapshot): Map<string, string> {
  return new Map(snapshot.teams.map((t) => [t.team_id, t.name]));
}
